### PR TITLE
Refactor interop server

### PR DESCRIPTION
**Pull request by johanstokking**
#### Summary

Main part of #4677 
Addresses #2627 (interop server and testing)
Addresses #2798 (`joinserver.proto`)
References #717 
References #718 

#### Changes

- Refactor `pkg/interop` server part
- Introduce more generic authorizer that replaces implementation in `pkg/joinserver`. This prepares for usage in IS (#4678) and allows for multiple authentication methods (like Packet Broker #4703)
- Resolved quite some technical debt: Echo dependency and gogoproto tags in `joinserver.proto`
- Support DNS SANs in client certificates instead of using CommonName for the origin address. Regenerated test certs to test this
- Interpret Network Server address as `HNSID` in `HomeNSAns`
- Remove other usages of NSID; it's 1.1 only and we have yet to see to which extent we're going to use `SenderNSID` and `ReceiverNSID` here
- Send correct message type depending on the protocol version, also some validation for incoming requests
- Cleaned up test cases

#### Testing

CI covers all logic. Also local integration testing:

```bash
$ mkdir -p .dev/interop
$ CAROOT=.dev/interop mkcert -cert-file .dev/interop/cert.pem -key-file .dev/interop/key.pem -client localhost "*.localhost"
$ echo "000001: rootCA.pem" > .dev/interop/config.yml
```

Configure TTS with:

```yaml
interop:
  source: directory
  directory: .dev/interop
```

Example `HomeNSReq` (notice `HNSID` that is set to `network_server_address`):

```bash
$ curl --cacert ca.pem --cert .dev/interop/cert.pem --key .dev/interop/key.pem -H "Content-Type: application/json" -d '{"ProtocolVersion":"1.1","MessageType":"HomeNSReq","SenderID":"000001","ReceiverID":"70B3D57ED0000000","DevEUI":"1122334455667788"}' https://localhost:8886

{"ProtocolVersion":"1.1","TransactionID":0,"MessageType":"HomeNSAns","SenderID":"70B3D57ED0000000","ReceiverID":"000001","Result":{"ResultCode":"Success"},"HNSID":"localhost","HNetID":"000001"}
```

Example `JoinReq`:

```
$ curl --cacert ca.pem --cert .dev/interop/cert.pem --key .dev/interop/key.pem -H "Content-Type: application/json" -d '{"ProtocolVersion":"1.0","MessageType":"JoinReq","SenderID":"000001","ReceiverID":"70B3D57ED0000000","DevEUI":"1122334455667788","TransactionID":42,"MACVersion":"1.0.3","PHYPayload":"00000000d07ed5b37088776655443322111100e6ed0bb7","DevAddr":"26ABCDEF","DLSettings":"FF","RxDelay":5,"CFList":"FFFFFFFFFFFFFFFFFFFFFFFFFFFFFF01"}' https://localhost:8886/

{"ProtocolVersion":"1.0","TransactionID":42,"MessageType":"JoinAns","SenderID":"70B3D57ED0000000","ReceiverID":"000001","PHYPayload":"205697687B6E09FC8FB45D3CA4A8B4C740D05DF36B6B7634D37A356556798D9712","Result":{"ResultCode":"Success"},"Lifetime":0,"NwkSKey":{"KEKLabel":"","AESKey":"9D9F8AFDD16C43975580831A92DAAFB3"},"AppSKey":{"KEKLabel":"","AESKey":"D958DC9AC061A4458C6422A5C531E815"},"SessionKeyID":"017C4BE9138646EB4A4C6F3A1853E28D"}
```

##### Regressions

None expected

#### Notes for Reviewers

Since it's effectively a new implementation of the server part of `pkg/interop`, the diff view isn't very useful. You might want to check out the code, reset the index to the target branch and see in your code editor where the changes are and if the current code makes sense.

Note that there are no major changes in testing so the existing tests act as regression test.

@adriansmares is the main reviewer.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/interop/server_errors.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	}
+
+	msg := ErrorMessage{
+		MessageHeader: answerHeader,
+		Result: Result{
+			ResultCode:  code,
+			Description: desc,
+		},
+	}
+	json.NewEncoder(w).Encode(msg)
```
Comment (adriansmares) [inline pkg/interop/server_errors.go:53]: Nit: `w.WriteHeader(http.StatusBadRequest)` on non-`nil` error. At least we get infrastructure errors if this happens.

File: pkg/joinserver/http_interop.go [hunk 2]
```diff
@@ -119,26 +124,40 @@ func (srv interopServer) JoinRequest(ctx context.Context, in *interop.JoinReq) (
 func (srv interopServer) HomeNSRequest(ctx context.Context, in *interop.HomeNSReq) (*interop.HomeNSAns, error) {
 	ctx = log.NewContextWithField(ctx, "namespace", "joinserver/interop")
 
-	netID, err := srv.JS.GetHomeNetID(ctx, types.EUI64(in.ReceiverID), types.EUI64(in.DevEUI), X509DNAuthorizer)
+	netID, nsID, err := srv.JS.GetHomeNetID(ctx, types.EUI64(in.ReceiverID), types.EUI64(in.DevEUI), InteropAuthorizer)
 	if err != nil {
+		switch {
+		case errors.Resemble(err, errRegistryOperation):
+			if errors.IsNotFound(errors.Cause(err)) {
+				return nil, interop.ErrUnknownDevEUI.WithCause(err)
```
Comment (adriansmares) [inline pkg/joinserver/http_interop.go:132]: Could also be `ReceiverID` / `JoinEUI` - maybe unknown EUIs ?

**Review by adriansmares — APPROVED** [review]
Did you find out what was causing the not found errors during our testing of `HomeNSReq` ?

File: pkg/interop/server_errors.go (hunk 1 above)
Comment (johanstokking) [reply pkg/interop/server_errors.go:53]: The headers are already sent as soon as the encoder emitted the first characters. I don't think we can do much here, other than first encoding the response in memory and then writing it. TBH, I don't see a case here where JSON marshaling fails; these structs are under our control, there are no functions being marshaled or `interface{}` fields etc

**Comment by johanstokking** [comment]
@adriansmares I appended a few commits, you can only look at those.

File: pkg/joinserver/http_interop.go (hunk 2 above)
Comment (johanstokking) [reply pkg/joinserver/http_interop.go:132]: Yes, in The Things Stack only the combination of JoinEUI and DevEUI is supposed to be unique. Generally speaking, LoRaWAN Network Servers keep DevEUIs unique, regardless of the JoinEUI.

Therefore there is no error code in Backend Interfaces that says `UnknownEUIs`; it's either `UnkonwnDevEUI` or `UnknownReceiver`.

Now, `UnkownReceiver` (sic) should be returned when the `ReceiverID` isn't known, i.e. when you're contacting a Join Server that is not handling this JoinEUI. We might want to support this through config (like a list of JoinEUI prefixes), but I don't really see a need for it.

**Comment by johanstokking** [comment]
> Did you find out what was causing the not found errors during our testing of `HomeNSReq` ?

No, but things changed dramatically already. I didn't put time in figuring out what the reason was in the current version. It's probably something really stupid. I did some integration testing with this code to see if it really is resolved and it looks like it.

File: pkg/interop/messages.go [hunk 3]
```diff
@@ -58,6 +58,18 @@ type ErrorMessage struct {
 	Result Result
 }
 
+// NsMessageHeader contains the message header for NS messages.
+type NsMessageHeader struct {
+	MessageHeader
+	SenderID   NetID
```
Comment (adriansmares) [inline pkg/interop/messages.go:64]: Is the `0x` handling supposed to be removed ? I expected it to be in the unmarshaller maybe.

Comment (johanstokking) [reply pkg/interop/messages.go:64]: Yeah it's there, it goes via `Buffer`: https://github.com/TheThingsNetwork/lorawan-stack/blob/feature/4677-interop/pkg/interop/types.go#L279-L287 and https://github.com/TheThingsNetwork/lorawan-stack/blob/26e787cdd48bf43ce7372d6a48a1a92bc35aebb7/pkg/interop/types.go#L345-L348

**Review by adriansmares — APPROVED** [review]

File: pkg/config/shared.go [hunk 4]
```diff
@@ -414,6 +414,7 @@ func (c SenderClientCA) Fetcher(ctx context.Context) (fetch.Interface, error) {
 // InteropServer represents the server-side interoperability through LoRaWAN Backend Interfaces configuration.
 type InteropServer struct {
 	ListenTLS                string            `name:"listen-tls" description:"Address for the interop server to listen on"`
+	TrustedProxies           []string          `name:"trusted-proxies" description:"CIDRs of trusted reverse proxies"`
```
Comment (htdvisser) [inline pkg/config/shared.go:417]: This should be mentioned in CHANGELOG.

**Review by htdvisser — APPROVED** [review]
Looks good to me 👍

**Pull request merged**